File: hedging/reservation-service/src/main/java/com/example/reservationservice/ReservationServiceApplication.java
package com.example.reservationservice;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.log4j.Log4j2;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.annotation.Bean;
import org.springframework.context.event.EventListener;
import org.springframework.core.env.Environment;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.repository.ReactiveMongoRepository;
import org.springframework.stereotype.Component;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;
import reactor.core.publisher.Flux;

import static org.springframework.web.reactive.function.server.RequestPredicates.GET;
import static org.springframework.web.reactive.function.server.RouterFunctions.route;
import static org.springframework.web.reactive.function.server.ServerResponse.ok;

@SpringBootApplication
public class ReservationServiceApplication {

	@Bean
	RouterFunction<ServerResponse> routes(ReservationRepository rr,
																																							Environment env) {
		return
			route(GET("/reservations"), r -> ok().body(rr.findAll(), Reservation.class))
				.andRoute(GET("/message"), request -> ServerResponse.ok().syncBody(env.getProperty("message")));
	}

	public static void main(String[] args) {
		SpringApplication.run(ReservationServiceApplication.class, args);
	}
}

@Log4j2
@Component
class Initializer {

	private final ReservationRepository reservationRepository;

	Initializer(ReservationRepository reservationRepository) {
		this.reservationRepository = reservationRepository;
	}

	@EventListener(ApplicationReadyEvent.class)
	public void listen() throws Exception {

		Flux<Reservation> saved = Flux.just("A", "B", "C")
			.map(nom -> new Reservation(null, nom))
			.flatMap(this.reservationRepository::save);

		this.reservationRepository
			.deleteAll()
			.thenMany(saved)
			.thenMany(this.reservationRepository.findAll())
			.subscribe(log::info);
	}
}


interface ReservationRepository extends ReactiveMongoRepository<Reservation, String> {
}

@Document
@Data
@AllArgsConstructor
@NoArgsConstructor
class Reservation {
	@Id
	private String id;
	private String reservationName;
}

File: finchley/reservation-service-kotlin/src/main/kotlin/com/example/reservationservice/ReservationServiceApplication.kt
package com.example.reservationservice

import org.springframework.boot.ApplicationRunner
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.cloud.stream.annotation.EnableBinding
import org.springframework.cloud.stream.annotation.Input
import org.springframework.cloud.stream.annotation.StreamListener
import org.springframework.cloud.stream.messaging.Sink
import org.springframework.context.annotation.Configuration
import org.springframework.context.support.beans
import org.springframework.core.env.Environment
import org.springframework.core.env.get
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.repository.ReactiveMongoRepository
import org.springframework.web.reactive.function.server.ServerResponse.ok
import org.springframework.web.reactive.function.server.body
import org.springframework.web.reactive.function.server.router
import reactor.core.publisher.Flux

@SpringBootApplication
class ReservationServiceApplication

@Configuration
@EnableBinding(Sink::class)
class Processor(val rr: ReservationRepository) {

	@StreamListener
	fun go(@Input(Sink.INPUT) incoming: Flux<String>) {
		incoming
				.map { it.toUpperCase() }
				.flatMap { rr.save(Reservation(reservationName = it)) }
				.subscribe()
	}
}

fun main(args: Array<String>) {
	SpringApplicationBuilder()
			.initializers(beans {
				bean {
					ApplicationRunner {
						val repo = ref<ReservationRepository>()
						val names = Flux.just("Mario", "Mia", "Tam Mie", "Josh")
								.map { Reservation(reservationName = it) }
								.flatMap { repo.save(it) }
						repo
								.deleteAll()
								.thenMany(names)
								.thenMany(repo.findAll())
								.subscribe { println(it) }
					}
				}
				bean {
					val ref = ref<ReservationRepository>()
					val env = ref<Environment>()
					router {
						GET("/message") { ok().body(Flux.just(env["message"])) }
						GET("/reservations") { ok().body(ref.findAll()) }
					}
				}
			})
			.sources(ReservationServiceApplication::class.java)
			.run(*args)
}

interface ReservationRepository : ReactiveMongoRepository<Reservation, String>

@Document
data class Reservation(@Id var id: String? = null, var reservationName: String? = null)